### PR TITLE
Enable session based authentication

### PR DESCRIPTION
**Pull request by romidane**
Implementation
- [x] Remove use of JWT headers for athentication
- [x] Refactor frontend application to not require knowledge of JWT
- [x] Create Server side longin page and remove the clientside react one
- [x] Add session cookie
- [x] Add hmpps-session-cookie which holds the JTW tokens
- [x] Update application routing
- [x] Update tests

Extras
- [x] Add security headers

### CODE REVIEW

File: backend/index.js [hunk 1]
```diff
@@ -17,11 +24,24 @@ const manualoverride = require('./controllers/manualoverride');
...
+
+const sixtyDaysInSeconds = 5184000;
+const sessionExpiryMinutes = config.session.expiryMinutes * 60 * 1000;
+
+const sessionConfig = {
+  name: config.session.name,
+  secret: config.session.secret,
+  sameSite: true,
+  expires: new Date(Date.now() + sessionExpiryMinutes),
+  maxAge: sessionExpiryMinutes // 1 hour
```
Comment (glenjamin) [inline backend/index.js]: this comment might age badly :)

File: backend/index.js [hunk 2]
```diff
@@ -34,30 +54,67 @@ function getAppInfo () {
...
-app.use('/unallocated', jsonParser, unallocated.router);
-app.use('/allocated', jsonParser, allocated.router);
-app.use('/userLocations', jsonParser, userLocations);
-app.use('/searchOffenders', jsonParser, searchOffenders.router);
-app.use('/manualoverride', jsonParser, manualoverride);
-app.use('/keyworkerSearch', jsonParser, keyworkerSearch);
-app.use('/keyworker', jsonParser, keyworkerProfile);
-app.use('/keyworkerAllocations', jsonParser, keyworkerAllocations.router);
+app.use(cookieSession(sessionConfig));
+app.use(cookieParser());
```
Comment (glenjamin) [inline backend/index.js]: I always thought `cookieParser` had to be before `cookieSession`?

File: backend/index.js [hunk 3]
```diff
@@ -34,30 +54,67 @@ function getAppInfo () {
...
+});
 
 app.use('/health', require('express-healthcheck')());
 
 app.use('/info', (req, res) => {
   res.json(getAppInfo());
 });
 
-app.use('/', express.static(path.join(__dirname, '../build')));
+app.use(express.static(path.join(__dirname, '../public'), { index: '_' }));
```
Comment (glenjamin) [inline backend/index.js]: maybe `"dummy-file-which-doesnt-exist"` and/or a comment to explain the `_`

File: backend/session.js [hunk 4]
```diff
@@ -1,35 +1,118 @@
...
+};
+
+const updateHmppsCookie = (res) => (tokens) => {
+  setHmppsCookie(res, tokens);
 };
 
-const isAuthenticated = (headers) => getSessionData(headers) !== null;
-const extendSession = (headers) => newJWT(getSessionData(headers));
+const deleteHmppsCookie = (res) => {
+  res.cookie(config.hmppsCookie.name, '', { expires: new Date(0), domain: config.hmppsCookie.domain, path: '/' });
```
Comment (glenjamin) [inline backend/session.js]: Are you sure this doesn't need all the other matching cookie config like `secure` and `httpOnly`?

I think i'd be happier if this shared the config object with the `set` code.

File: public/styles/app.css [hunk 5]
```diff
@@ -0,0 +1,140 @@
+.page-header {
```
Comment (glenjamin) [inline public/styles/app.css]: Now that we moved `../build` into the top of the app, do we still need to check in all this `public` stuff?

File: backend/index.js (hunk 1 above)
Comment (romidane) [reply backend/index.js]: Good catch 👍

File: backend/session.js (hunk 4 above)
Comment (romidane) [reply backend/session.js]: I tried add all the options but that didn't work.. turns out only the domain and expire are necessary for this

File: public/styles/app.css (hunk 5 above)
Comment (romidane) [reply public/styles/app.css]: For the login page we do need it .. the imported styles are being loaded as style tags in the html.. the requests to public will only send something for login I believe

Comment (glenjamin) [reply public/styles/app.css]: I think i'd rather we didn't have a second separate copy of the CSS kicking around, but i'm not particularly bothered either way.

The approach you started taking with reading the manifest to render the HTML and use the webpack built assets would acheive this.

**Pull request merged**